### PR TITLE
Group Data Provider updated to newest Group Key Management spec.

### PR DESCRIPTION
**Pull request by rcasallas-silabs**
#### Problem
The Group Data Provider lacks support for the new Group Key Management attributes.

#### Change overview
* Group Key Management interface updated:
   * GroupMapping functions replaced. The data provider should support two types of mappings, groups to endpoint, and groups to keysets.
   * Three new sections replace GroupMappings: GroupInfo, Endpoints, and GroupKeys.
   * Unit tests updated to support the new functions.
   * Groups cluster server updated to use the new interface.

#### Testing
* Unit tests run successfully (existing KeySet tests, plus new tests).
* Existing YAML group cluster tests executed successfully.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/credentials/GroupDataProviderImpl.h [hunk 1]
```diff
@@ -72,95 +79,83 @@ class GroupDataProviderImpl : public GroupDataProvider
...
     };
 
     chip::PersistentStorageDelegate & mStorage;
     bool mInitialized = false;
-    BitMapObjectPool<AllGroupMappingsIteratorImpl, kIteratorsMax> mAllGroupsIterators;
-    BitMapObjectPool<GroupMappingIteratorImpl, kIteratorsMax> mEndpointGroupsIterators;
-    BitMapObjectPool<AllStatesIterator, kIteratorsMax> mAllStatesIterators;
-    BitMapObjectPool<FabricStatesIterator, kIteratorsMax> mFabricStatesIterators;
-    BitMapObjectPool<KeySetIteratorImpl, kIteratorsMax> mKeySetIterators;
+    BitMapObjectPool<GroupInfoIteratorImpl, kIteratorsMax, OnObjectPoolDestruction::IgnoreUnsafeDoNotUseInNewCode>
```
Comment (kpschoedel) [inline src/credentials/GroupDataProviderImpl.h:152]: Don't do this; call `BitMapObjectPool::ReleaseAll()` in the enclosing class shutdown or destructor if there's no other management of the objects.

**Review by kpschoedel — CHANGES_REQUESTED** [review]

File: src/app/clusters/groups-server/groups-server.cpp [hunk 2]
```diff
@@ -151,97 +132,61 @@ void emberAfGroupsClusterServerInitCallback(EndpointId endpointId)
...
-        SuccessOrExit(err = commandObj->FinishCommand());
-    }
+    response.groupName = CharSpan(info.name, strnlen(info.name, GroupDataProvider::GroupInfo::kGroupNameMax));
+    status             = EMBER_ZCL_STATUS_SUCCESS;
 exit:
-    if (err != CHIP_NO_ERROR)
+    response.groupId = groupId;
+    response.status  = status;
+    err              = commandObj->AddResponseData(commandPath, response);
+    if (CHIP_NO_ERROR != err)
```
Comment (bzbarsky-apple) [inline src/app/clusters/groups-server/groups-server.cpp:186]: This does not look right to me.  If we got here with a failure status, we should be sending the status, not this response....

**Review by bzbarsky-apple — COMMENTED** [review]
Note: I did not review the GroupDataProviderImpl changes.

File: src/credentials/GroupDataProviderImpl.h (hunk 1 above)
Comment (rcasallas-silabs) [reply src/credentials/GroupDataProviderImpl.h]: ReleaseAll() is now called from the Finish() function.

File: src/app/clusters/groups-server/groups-server.cpp (hunk 2 above)
Comment (rcasallas-silabs) [reply src/app/clusters/groups-server/groups-server.cpp:186]: I believe this function follows the spec. According to section 1.3.6.2.1. whatever the status is, it should be included in the response message ViewGroupResponse, which contains a `status` field.

**Review by kpschoedel — APPROVED** [review]

File: src/app/clusters/groups-server/groups-server.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/clusters/groups-server/groups-server.cpp:186]: Ah, good point.  Thank you!

**Pull request merged**